WRR-976: ContextualPopupDecorator: Modified to update position when screen orientation change

mmyelyn opened the pull request:
### Checklist

* [X] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [x] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [X] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
When ContextualPopupDecorator or ContextualMenuDecorator is open, rotating the device in the inspector should change the popup's position. Currently the popup is fixed and does not update.


### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
I needed to check whether the position had not been updated or the value had been entered incorrectly. When rotated, I saw that when I refreshed, the position was calculated normally again. Therefore, the position calculation was no problem.

The position is updated with the positionContextualPopup function, which is not called when the window size changes, including when the screen is rotated. Therefore, I modified it to use ResizeObserver to detect changes in FloatLayer. 
(document.body may or may not have a size depending on the app)



### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)


### Links
[//]: # (Related issues, references)
WRR-976

### Comments
Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

MikyungKim commented on CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@ The following is a curated list of changes in the Enact sandstone module, newest
 ### Fixed
 
 - `sandstone/Steps` prop `size` to accept number type
+- `sandstone/ContextualPopupDecorator` to re-position when window size is changed
```
```suggestion
- `sandstone/ContextualPopupDecorator` to update its position when the window size has been resized
```

MikyungKim commented on ContextualPopupDecorator/tests/ContextualPopupDecorator-specs.js:
```diff
@@ -505,4 +505,35 @@ describe('ContextualPopupDecorator Specs', () => {
 		expect(scrimDivSecond).toHaveClass(expectedSecond);
 	});
+
+	test('should re-positino when change window size', () => {
```
```suggestion
	test('should update its position when the window size has been changed', () => {
```

MikyungKim commented on ContextualPopupDecorator/ContextualPopupDecorator.js:
```diff
@@ -292,6 +292,7 @@ const Decorator = hoc(defaultConfig, (config, Wrapped) => {
 				on('keydown', this.handleKeyDown);
 				on('keyup', this.handleKeyUp);
 			}
+			window.addEventListener('resize', this.handleResize);
```
Let's check the existence of window.

```suggestion
			if (typeof window !== 'undefined') {
				window.addEventListener('resize', this.handleResize);
			}
```

seunghoh replied:
@mmyelyn @vJIYEv Please keep in mind that you should always check `window` for isomorphic + v8 snapshot build in Enact

MikyungKim commented on ContextualPopupDecorator/ContextualPopupDecorator.js:
```diff
@@ -341,6 +342,7 @@ const Decorator = hoc(defaultConfig, (config, Wrapped) => {
 				off('keyup', this.handleKeyUp);
 			}
 			Spotlight.remove(this.state.containerId);
+			window.removeEventListener('resize', this.handleResize);
```
```suggestion
			if (typeof window !== 'undefined') {
				window.removeEventListener('resize', this.handleResize);
			}
```

MikyungKim commented:
@mmyelyn , Please update the checklist of the PR description.

MikyungKim reviewed (commented):
![image](https://github.com/user-attachments/assets/31b6ee0b-1773-481c-a474-39868bcdb3d4)
Sometimes the changed activator's metric is not updated yet when the `resize` event triggered.
Maybe it's better to use `ResizeObserver` instead of `resize` event.
Please refer to `useScroll` case. I guess we need to observe `this.clientSiblingRef` metrics.

vJIYEv commented:
Period : 2024/08/31 00:00 ~ 2024/09/06 23:55 [Korea Standard Time]

부재중입니다.

mmyelyn commented:
@MikyungKim I changed to handle resizeObserver instead of calling resize handler. but, When observing `this.clientSiblingRef`, the size of the contextual button does not change, so I added document.body is observed. 

The div where the size exists is inside a floatingLayer component, so I couldn't make it observe this.

MikyungKim commented on ContextualPopupDecorator/ContextualPopupDecorator.js:
```diff
@@ -292,6 +295,16 @@ const Decorator = hoc(defaultConfig, (config, Wrapped) => {
+				});
+			}
+
+			if (this.clientSiblingRef?.current) {
+				this.resizeObserver.observe(document.body);
+			}
```
We'd better move this to the below timing since it will recalculate the position even the contexualpopup is closed.
Also, it's better to observe the real DOM that has size below "floatLayer" since we cannot gurantee that body has always its size.
We need to discuss better selectors if any.

```
getContainerNode = (node) => {
			this.containerNode = node;

			if (this.resizeObserver && typeof document !== 'undefined') {
				if (node) {
					this.resizeObserver.observe(document.querySelector(`#floatLayer > div`));
				} else {
					this.resizeObserver.disconnect();
				}
			}
		};
```

mmyelyn replied:
That's a good point. This is the part I missed, and I reflected your reviews

MikyungKim commented on ContextualPopupDecorator/tests/ContextualPopupDecorator-specs.js:
```diff
@@ -505,4 +514,20 @@ describe('ContextualPopupDecorator Specs', () => {
 		expect(scrimDivSecond).toHaveClass(expectedSecond);
 	});
+
+	test('should creates and observes with ResizeObserver', () => {
```
```suggestion
	test('should create and observe with `ResizeObserver`', () => {
```

MikyungKim commented on CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@ The following is a curated list of changes in the Enact sandstone module, newest
 ### Fixed
 
 - `sandstone/Steps` prop `size` to accept number type
+- `sandstone/ContextualPopupDecorator` to update its position when the window size has been resized
```
Please sync your branch with the latest develop.
And put the change log in alphabetical order under `Fixed` category.

mmyelyn replied:
Thanks, Modified

MikyungKim commented on ContextualPopupDecorator/tests/ContextualPopupDecorator-specs.js:
```diff
@@ -25,6 +27,13 @@ describe('ContextualPopupDecorator Specs', () => {
 			};
 		});
+
+		MockObserverInstance = {
```
It seems we don't need this code anymore.

MikyungKim commented on ContextualMenuDecorator/tests/ContextualMenuDecorator-specs.js:
```diff
@@ -10,7 +10,18 @@ const ContextualButton = ContextualMenuDecorator(Button);
+let MockObserverInstance;
+
 describe('ContextualMenuDecorator Specs', () => {
+	beforeEach(() => {
```
It seems we don't need this code anymore.

mmyelyn replied:
There is code for the existing test in beforeEach. Except that, the resizeObserver implementation has been moved to that test.

MikyungKim commented on Dropdown/tests/Dropdown-specs.js:
```diff
@@ -12,7 +12,18 @@ const children = ['option1', 'option2', 'option3'];
+let MockObserverInstance;
+
 describe('Dropdown', () => {
+	beforeEach(() => {
```
It seems we don't need this code anymore.

MikyungKim commented on ContextualPopupDecorator/tests/ContextualPopupDecorator-specs.js:
```diff
@@ -505,4 +505,40 @@ describe('ContextualPopupDecorator Specs', () => {
 		expect(scrimDivSecond).toHaveClass(expectedSecond);
 	});
+
+	test('should create and observe with `ResizeObserver` and disconnect when the popup close', () => {
```
```suggestion
	test('should create and observe with `ResizeObserver` when the popup opened and disconnect when the popup closed', () => {
```

MikyungKim commented:
@mmyelyn , Would you resolve the merge conflict?

MikyungKim reviewed (approved):
LGTM!

MikyungKim commented:
@mmyelyn, Could you please resolve the merge conflict?
